river/drift/adwin_c.pyx

```diff
@@ -0,0 +1,359 @@
+        exit_flag = False
+        self.tick += 1
+
+        # Reduce window
+        if (self.tick % self.clock == 0) and (self.width > self.grace_period):
+            reduce_width = True
+            while reduce_width:
+                reduce_width = False
+                exit_flag = False
+                n0 = 0.0            # length of window 0
+                n1 = self.width     # length of window 1
+                u0 = 0.0            # total of window 0
+                u1 = self.total     # total of window 1
+                v0 = 0              # variance of window 0
+                v1 = self.variance  # variance of window 1
+
+                # Evaluate each window cut (W_0, W_1)
+                for idx in range(len(self.bucket_deque) - 1, -1 , -1):
+                    if exit_flag:
+                        break
+                    bucket = self.bucket_deque[idx]
+
+                    for k in range(bucket.current_idx - 1):
+                        n2 = self._calculate_bucket_size(idx)   # length of window 2  TODO confirm index
+                        u2 = bucket.get_total_at(k)             # total of window 2
+                        # Warning: means are calculated inside the loop to get updated values.
+                        mu2 = u2 / n2   # mean of window 2
+
+                        if n0 > 0.0:
+                            mu0 = u0 / n0  # mean of window 0
+                            v0 += (
+                                    bucket.get_variance_at(k) + n0 * n2
+                                    * (mu0 - mu2) * (mu0 - mu2)
+                                    / (n0 + n2)
+                            )
+
+                        if n1 > 0.0:
+                            mu1 = u1 / n1  # mean of window 1
+                            v1 -= (
+                                    bucket.get_variance_at(k) + n1 * n2
+                                    * (mu1 - mu2) * (mu1 - mu2)
+                                    / (n1 + n2)
+                            )
+
+                        # Update window 0 and 1
+                        n0 += self._calculate_bucket_size(idx)
+                        n1 -= self._calculate_bucket_size(idx)
+                        u0 += bucket.get_total_at(k)
+                        u1 -= bucket.get_total_at(k)
+
+                        if (idx == 0) and (k == bucket.current_idx - 1):
+                            exit_flag = True    # We are done
+                            break
+
+                        # Check if delta_mean < epsilon_cut holds
+                        # Note: Must re-calculate means per updated values
+                        delta_mean = (u0 / n0) - (u1 / n1)
+                        if (
+                                n1 >= <double> self.min_window_length
+                                and n0 >= <double> self.min_window_length
+                                and self._evaluate_cut(n0, n1, delta_mean, self.delta)
+                        ):
```

Comment:
Is there a reason to cast to double inside the if? (possible noob question)

Reply:
Not really, it was left from debugging. I have removed it.